### PR TITLE
fix redundant_pattern_matching lint

### PR DESCRIPTION
**Pull request by alex-700**
- now it handles `while let` case  (related to #5462)
- better suggestions in `if let` case

changelog: Fix suggestion in `redundant_pattern_matching` and also apply this lint to the `while let` case

### CODE REVIEW

**Comment by alex-700** [comment]
And I have a couple questions. (so PR is WIP for now)

- this lint for `while let` "conflicts" with the [while-let-on-iterator](https://rust-lang.github.io/rust-clippy/master/index.html#while_let_on_iterator) lint. What should we do with this? It seems not okay to give out two lints about on piece of code.

- before this PR the most of suggestions from this lint was completely wrong. It seems that we need replace only `while/if let PATS = EXPR` part, but not whole expression/statement. But this way can generate a code like `if o.is_some() { true } else { false }` that should be fix by another lint. What should we do with this?

Thanks.

**Comment by flip1995** [comment]
- This is kind of a problem, but can be fixed quite easily. You can just copy this check https://github.com/rust-lang/rust-clippy/blob/52dacbc876f3f60f25cdbdf0e8a0a3737beb6704/clippy_lints/src/loops.rs#L570-L571 to this lint and return if the `match_expr` is a `Iterator::next` method call. `match_expr` is in this case `op` and the `method_path` you can get by matching `if let ExprKind::MethodCall(method_path) = op.kind`.

- Thanks for fixing this suggestion! When a suggestion leads to another lint triggering, it is not a problem as long as it doesn't become a cycling suggestion, which is not the case here.

File: clippy_lints/src/redundant_pattern_matching.rs [hunk 1]
```diff
@@ -86,19 +85,20 @@ fn find_sugg_for_if_let<'a, 'tcx>(
...
+            // in the case of WhileLetDesugar expr.span == op.span incorrectly.
+            // this is a workaround to restore true value of expr.span
+            let expr_span = expr.span.to(arms[1].span);
+            let span = expr_span.until(op.span.shrink_to_hi());
             diag.span_suggestion(
                 span,
                 "try this",
-                format!("{}.{}{}", snippet(cx, op.span, "_"), good_method, maybe_semi),
+                format!("{} {}.{}", keyword, snippet(cx, op.span, "_"), good_method),
                 Applicability::MaybeIncorrect, // snippet
```
Comment (flip1995) [inline clippy_lints/src/redundant_pattern_matching.rs:102]: I think this is
```suggestion
                Applicability::MachineApplicable, // snippet
```
now

**Comment by alex-700** [comment]
@flip1995 
- I'll do this, but it looks strange a bit. Is it a usual situation to check that we don't trigger lint in the code of the other lint? Maybe is there a better machinery for it?
- Is it ok to allow some "triggered after fixit" lints in tests (as I did here)?

**Comment by flip1995** [comment]
- Lints overlapping is pretty unusual, so we don't have anything to prevent this. Normally, if lints overlap, they are pretty similar and therefore are implemented in the same module, where it is easier to separate them ("if _a_ trigger lint 1 if _b_ trigger lint 2") I think just introducing this check here (with a comment why it is there) is the easiest way to do this in this case.

- Yes, that is the standard approach for this.

File: clippy_lints/src/redundant_pattern_matching.rs (hunk 1 above)
Comment (alex-700) [reply clippy_lints/src/redundant_pattern_matching.rs]: Agree. Fixed.

**Comment by alex-700** [comment]
@flip1995 sorry, should I fix something here? Changelog? What should be written in changelog?

And I think that I fix #4344 too.

**Comment by flip1995** [comment]
Don't worry about the changelog, this only means, that the PR body must contain a line `changelog: ...` which describes, what this PR does in one sentence. I added that for you.

@bors retry

**Pull request merged**